---
title: "DADA2 processing of Db-MM amplicon sequencing"
author: "Sudarshan A. Shetty"
date: "`r date()`"
output: 
  html_document: 
    toc: yes
    toc_depth: 2
    toc_float: true
    code_folding: hide  
editor_options: 
  chunk_output_type: console
---


# Diet based minimal microbiome (Db-MM-10)study 16S rRNA analysis

## load libraries  

```{r, message=FALSE, warning=FALSE, eval=FALSE}
#library("devtools"); 
#devtools::install_github("timelyportfolio/sankeytree")

library(DT) 
library(ggplot2); packageVersion("ggplot2")
library(dada2); packageVersion("dada2")
library(dplyr)
library(microbiome)
library(microbiomeutilities)
library(phangorn)
library(ggtree)
library(DECIPHER)
#library(ggrepel)
```


Copied the chunck from DADA2 tutorial and modified to this specific analysis.  

## File parsing

The fastq files were demultiplex using NG_tax galaxy version and stored in `data/fastq` folder.  
```{r eval=FALSE}

# File parsing
pathF <- "data/fastq/fwd/" # CHANGE ME to the directory containing your demultiplexed forward-read fastqs
pathR <- "data/fastq/rev/" # CHANGE ME ...
filtpathF <- file.path(pathF, "filtered") # Filtered forward files go into the pathF/filtered/ subdirectory
filtpathR <- file.path(pathR, "filtered") # ...

fastqFs <- sort(list.files(pathF, pattern=".fastq"))
fastqRs <- sort(list.files(pathR, pattern=".fastq"))


if(length(fastqFs) != length(fastqRs)) stop("Forward and reverse files do not match.")
# Filtering: THESE PARAMETERS ARENT OPTIMAL FOR ALL DATASETS
filterAndTrim(fwd=file.path(pathF, fastqFs), filt=file.path(filtpathF, fastqFs),
              rev=file.path(pathR, fastqRs), filt.rev=file.path(filtpathR, fastqRs),
              truncLen=c(90,90), maxEE=2, truncQ=2, maxN=0, rm.phix=TRUE,
              compress=TRUE, verbose=TRUE, multithread=F)
```


```{r eval=FALSE}
# File parsing
filtpathF <- "data/fastq/fwd/filtered" # CHANGE ME to the directory containing your filtered forward fastqs
filtpathR <- "data/fastq/rev/filtered" # CHANGE ME ...
filtFs <- list.files(filtpathF, pattern="fastq", full.names = TRUE)
filtRs <- list.files(filtpathR, pattern="fastq", full.names = TRUE)


sample.names <- sapply(strsplit(basename(filtFs), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz
sample.namesR <- sapply(strsplit(basename(filtRs), "_"), `[`, 1) # Assumes filename = samplename_XXX.fastq.gz

if(!identical(sample.names, sample.namesR)) stop("Forward and reverse files do not match.")
names(filtFs) <- sample.names
names(filtRs) <- sample.names
set.seed(19883)
# Learn forward error rates
# SOME ISSUE with MERGING so GOING FOR Forwards only
#errF <- learnErrors(filtFs, nbases=1e8, multithread=F, randomize=TRUE)
# Learn reverse error rates
#errR <- learnErrors(filtRs, nbases=1e8, multithread=F, randomize=TRUE)

err <- learnErrors(filtFs, nbases = 1e8, multithread=F, randomize=TRUE)
# Infer sequence variants
dds <- vector("list", length(sample.names))
dds <- vector("list", length(sample.names))
names(dds) <- sample.names

for(sam in sample.names) {
  cat("Processing:", sam, "\n")
  derep <- derepFastq(filtFs[[sam]])
  dds[[sam]] <- dada(derep, err=err, multithread=TRUE)
}

head(dds)
# Construct sequence table and write to disk

seqtab <- makeSequenceTable(dds)

saveRDS(seqtab, "output/rds/seqtab.rds")

```

## Remove chimera and assign taxonomy  

For taxonomy a custome file was created to included only 16S sequecing from species that are part of Db-MM.  

```{r eval=FALSE}
set.seed(55536)
seqtab2 <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE)
# Assign taxonomy
dbmmdb <- "data/rrna_MMDb/dbmm_tax_db.fas"
tax <- assignTaxonomy(seqtab2, dbmmdb, 
                      minBoot=50, verbose = T,
                     outputBootstraps = F)

# Write to disk
saveRDS(tax, "output/rds/tax_final.rds") # CHANGE ME ...
saveRDS(seqtab2, "output/rds/seqtab2.rds") 
```


## Build tree  
```{r eval=FALSE}
# maybe avoid this step
library(DECIPHER)
library(phangorn)
unique(tax[,"Genus"])

seqs <- getSequences(seqtab2)
names(seqs) <- seqs # This propagates to the tip labels of the tree
alignment <- AlignSeqs(DNAStringSet(seqs), anchor=NA)

phang.align <- phyDat(as(alignment, "matrix"), type="DNA")
dm <- dist.ml(phang.align)
treeNJ <- NJ(dm) # Note, tip order != sequence order
fit = pml(treeNJ, data=phang.align)

## negative edges length changed to 0!

fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE,
                      rearrangement = "stochastic", control = pml.control(trace = 0))
detach("package:phangorn", unload=TRUE)

saveRDS(fitGTR, "output/rds/fitGTR_tree.rds") 

```

## Make phyloseq object 

```{r eval=FALSE}
# read metadata table
metatable <- read.csv("data/map_dbmm.csv", stringsAsFactors = F, row.names = 1)
samdat <- sample_data(sample_data(metatable))
rownames(samdat)
rownames(seqtab2)
# make phyloseq object
pseq01 <- phyloseq(otu_table(t(seqtab2), taxa_are_rows=T), sample_data(samdat),
               tax_table(tax))

saveRDS(pseq01, "output/rds/pseq01.rds") 

```

The phyloseq object pseq01 is the input for 02_post_dada2_qmp_analysis.  


```{r}
sessionInfo()
```


